### DISCUSSION TITLE
Welcome to WTF Dial Discussions

### DISCUSSION THREAD

**Discussion by benbjohnson** [benbjohnson/wtf#1] [General]
## Welcome! 👋

WTF Dial is an example project for how to structure real-world Go applications. You can find companion blog posts that explain the architecture and design at [Go Beyond](https://www.gobeyond.dev/). Instead of using a traditional comments system, I wanted to try GitHub Discussions instead.

I'd love to hear what you're interested to learn and what you're building yourself.  Please comment below and introduce yourself to the community.

**Comment by komuw** [comment]
Hi, 
I'm Komu. The last few months I built https://github.com/komuw/kama which prints out a types documentation. 

One question I had while perusing through the `wtf` code, there's hardly any logging in the code; is that by design?   
I noticed that the code has metrics via prometheus but no logging.

**Reply by benbjohnson** [reply]
Hi Komu. 👋

I tend to use logging in two instances:

1. Just before displaying an error to the end-user—I do this with the `Error()` function in the `http` package.
2. Long-running goroutines that have no other way to report errors. Unfortunately, there are not any background goroutines in `wtf` so there aren't any examples of this.

I find metrics are more useful than logs for tracking HTTP requests. It also keeps the log file only focused on errors and not informational messages.

**Reply by peterbourgon** [reply]
Seconding @benbjohnson — I've also basically stopped logging in my services, except for config details at startup, and critical errors. I find that metrics provide more value at lower cost, and when I need more information for e.g. troubleshooting, I tend to reach for request tracing — not distributed tracing like Zipkin, but in-process traces and annotations like x/net/trace — or, increasingly, profiles.

**Reply by flowchartsman** [reply]
As far as profiling and metrics, has anyone experimented with continuous profiling/tracing (e.g. [profefe](https://github.com/profefe/profefe))? It seems like a good way to bridge the gap between metrics and manual profiling.

**Reply by benbjohnson** [reply]
I haven't used `profefe` myself but it sounds awesome.

**Reply by peterbourgon** [reply]
I personally don't understand the value proposition for continuous profiling, but maybe I just haven't seen the light?

**Reply by benbjohnson** [reply]
I can the value in situations where you have infrequent & difficult to reproduce performance issues. It'd be great to go back retroactively and see profiles from a point-in-time.

**Reply by YoRHa-A5** [reply]
Really want to make it work for my project, but I guess I bit of more than I can chew right now trying to use a different DB that is completely new to me and at the same time trying to introduce Gin with this while still learning the ins & outs of the language 😅
